crates/matrix-sdk/src/sliding_sync/mod.rs

```diff
@@ -640,8 +644,21 @@ use crate::{config::RequestConfig, Client, Result};
 const MAXIMUM_SLIDING_SYNC_SESSION_EXPIRATION: u8 = 3;
 
 /// The Sliding Sync instance.
+///
+/// It is OK to clone this type as much as you need: cloning it is cheap.
 #[derive(Clone, Debug)]
 pub struct SlidingSync {
+    /// The Sliding Sync data.
+    inner: Arc<SlidingSyncInner>,
+
+    /// A lock to ensure that responses are handled one at a time.
+    /// [The lock][AsyncMutex] is fair, and this fairness property is important
+    /// to ensure responses are handled in the correct order.
```

Comment:
I don't think this is true. Responses will be handled in the order they arrive, which is not necessarily the order in which requests were made (although in practice it will be the vast majority of the time). The mutex has no effect on this at all except in the extremely unlikely case that **three or more** responses arrive in such a short amount of time that two response handling tasks are waiting to acquire the mutex at the same time while it is held by another task that was started first.

Reply:
That's correct, responses will be handled by “their arrival order”, not “request order”. I think it's a problem to tackle in another PR.

Reply:
Let's address that in another PR :-).